### DISCUSSION TITLE
How creator's non-SMBX 1.3 custom assets will be named without conflict for future developments

### DISCUSSION THREAD

**Discussion by P-Star7** [Wohlstand/TheXTech#118] [Ideas]
In future, TheXTech may allow for new background, background2, block, effect, level, npc, path, scene, and tiles. Users like me and Love Bodhi have desired more BGOs and world map sceneries per-level or etc. How will we allow the player to name these so that they can be recognized by TheXTech without download of config pack, or Wohlstand adds objects with IDs that the player already used?

My suggestion: player's customs are named such as "block-custom-1.png" or "block-1-custom.png". TheXTech will read these per-level without require a config pack or download, and Wohlstand's new hardcoded object additions will never conflict with these names. I think max array of each of these should be generous such as 1000 for the current time.

**Comment by ds-sloth** [comment]
I agree that allocating a unique namespace to custom assets is a good idea. To avoid clutter, maybe they can be "block-c1.png" instead of "block-custom-1.png"?

| I think max array of each of these should be generous such as 1000 for the current time.
We should also work to be careful with how much RAM these require. On low-RAM targets, a number of expanded maximums (especially `maxSections`) need to be returned to their original values to avoid stack-overflows. Perhaps the maximum number of new custom assets could follow the same `LOW_MEM` compile-time flag I have introduced.

**Reply by ChristianSilvermoon** [reply]
> I agree that allocating a unique namespace to custom assets is a good idea. To avoid clutter, maybe they can be "block-c1.png" instead of "block-custom-1.png"?
> 
> | I think max array of each of these should be generous such as 1000 for the current time.
> We should also work to be careful with how much RAM these require. On low-RAM targets, a number of expanded maximums (especially `maxSections`) need to be returned to their original values to avoid stack-overflows. Perhaps the maximum number of new custom assets could follow the same `LOW_MEM` compile-time flag I have introduced.

I'd think it should be possible to determine an episodes maximum memory usage at runtime. If determining this information is time consuming, the result of this 'max memory scan' should be stored in a metadata file and refreshed upon user request (if say, the episode has been edited).

If the episode comes out to an amount of memory usage that's too great for the system turn the Episode's name red on the Episode Select screen. When viewing the episode's saves display an "Insufficient Memory" message and do not allow any files to actually be started.

Alternatively if limits are exceeded (based on flags set at compile-time), then if an episode uses these custom things that aren't allowed, similarly should be disabled from being started so as not to allow you to play what doesn't work.

**Comment by Wohlstand** [comment]
If that a discussion, then I'll make it as discussion :wink: :fox_face:

**Comment by LoveBodhi** [comment]
"Love Bodhi", That's my alt Nickname. My main nickname is Yingchun Soul.

If there's The X-Tech intergration PGE Config Pack (CFGP in short, I don't use CP since that stands for **C**ou**p**ling), that was fine......

For Non-SMBX 1.3 / Advanced Asset (FTW for example): or use background-1c.png (like Editor-exclusive background-1e.png)?

**Comment by Wohlstand** [comment]
On my side the "TheXTech SDK" config pack in development, I gonna present it as replacement for "SMBX Integration" that no longer has reasons to be supported

**Comment by LoveBodhi** [comment]
I think my asset require to tweak in the future?

**Comment by ChristianSilvermoon** [comment]
I think #15 could be the answer to this discussion

EDIT: wait no, but would help along side it.

**Comment by LoveBodhi** [comment]
Unlike SMBX2, there're Init Ranges (751 - 1000).

**Comment by LoveBodhi** [comment]
I think that might require script system (which is planned) for more customizable elements.

e.g. Block Functions, NPC Functions, etc...

Also, I was opted out SMBX2.
